agent/agent.go

```diff
@@ -1,901 +1,299 @@
-	// execution behavior.
-	endChan chan *apimodels.TaskEndDetail
-
-	// signalHandler is used to process signals received by the agent during execution.
-	signalHandler *SignalHandler
-
-	// heartbeater handles triggering heartbeats at the correct intervals, and
-	// raises a signal if too many heartbeats fail consecutively.
-	heartbeater *comm.HeartbeatTicker
-
-	// statsCollector handles sending vital host system stats at the correct
-	// intervals, to the API server.
+type taskContext struct {
+	currentCommand command.Command
+	logger         client.LoggerProducer
 	statsCollector *StatsCollector
-
-	// metrics collector collects and sends process and system
-	// stats/metrics to the API server. This data is collected
-	// using native-(gopsutil) via the logging package.
-	metricsCollector *metricsCollector
-
-	// logger handles all the logging (task, system, execution, local)
-	// by appending log messages for each type to the correct stream.
-	logger *comm.StreamLogger
-
-	// timeoutWatcher maintains a timer, and raises a signal if the running task
-	// does not produce output within a given time frame.
-	idleTimeoutWatcher *comm.TimeoutWatcher
-
-	// maxExecTimeoutWatcher maintains a timer, and raises a signal if the running task
-	// does not return within a given time frame.
-	maxExecTimeoutWatcher *comm.TimeoutWatcher
-
-	// APILogger is a slogger.Appender which sends log messages
-	// to the API server.
-	APILogger *comm.APILogger
-
-	// Holds the current command being executed by the agent.
-	currentCommand      model.PluginCommandConf
-	currentCommandMutex sync.RWMutex
-
-	// taskConfig holds the project, distro and task objects for the agent's
-	// assigned task.
-	taskConfig *model.TaskConfig
-
-	// Registry manages plugins available for the agent.
-	Registry plugin.Registry
-
-	// currentTaskDir holds the absolute path of the directory that the agent has
-	// created for executing the current task.
-	currentTaskDir      string
-	currentTaskDirMutex sync.RWMutex
-
-	// agent's runtime configuration options.
-	opts Options
+	task           client.TaskData
+	taskConfig     *model.TaskConfig
+	taskDirectory  string
 }
 
-// finishAndAwaitCleanup sends the returned TaskEndResponse and error
-// for processing by the main agent loop.
-func (agt *Agent) finishAndAwaitCleanup(status string) (*apimodels.EndTaskResponse, error) {
-	// Signal all background actions to stop. If HandleSignals is still running,
-	// this will cause it to return.
-	defer close(agt.signalHandler.stopBackgroundChan)
-	defer agt.cleanup(agt.GetCurrentTaskId())
-	var detail *apimodels.TaskEndDetail
-	select {
-	case detail = <-agt.endChan:
-	default:
-		// endChan will be empty if the task completed without error
-		detail = agt.getTaskEndDetail()
+// New creates a new Agent with some Options and a client.Communicator. Call the
+// Agent's Start method to begin listening for tasks to run.
+func New(opts Options, comm client.Communicator) (*Agent, error) {
```

Comment:
I'm getting a compile error in tests because of this

./agent_integration_test.go:31: too many arguments to return
	have (*Agent, error)
	want (*Agent)
./status_test.go:55: multiple-value New() in single-value context

Reply:
done